src/components/safe-messages/MsgModal/index.tsx

```diff
@@ -0,0 +1,198 @@
+import RequiredIcon from '@/public/images/messages/required.svg'
+import { dispatchSafeMsgConfirmation, dispatchSafeMsgProposal } from '@/services/safe-messages/safeMsgSender'
+import useSafeInfo from '@/hooks/useSafeInfo'
+import { generateSafeMessageHash, generateSafeMessageMessage } from '@/utils/safe-messages'
+import { getDecodedMessage } from '@/components/safe-apps/utils'
+import useIsSafeOwner from '@/hooks/useIsSafeOwner'
+import useIsWrongChain from '@/hooks/useIsWrongChain'
+import ErrorMessage from '@/components/tx/ErrorMessage'
+import useAsync from '@/hooks/useAsync'
+import useWallet from '@/hooks/wallets/useWallet'
+import useSafeMessages from '@/hooks/useSafeMessages'
+import { isSafeMessageListItem } from '@/utils/safe-message-guards'
+import { useWeb3 } from '@/hooks/wallets/web3'
+
+import txStepperCss from '@/components/tx/TxStepper/styles.module.css'
+import { DecodedMsg } from '../DecodedMsg'
+import CopyButton from '@/components/common/CopyButton'
+
+const APP_LOGO_FALLBACK_IMAGE = '/images/apps/apps-icon.svg'
+const APP_NAME_FALLBACK = 'Sign message off-chain'
+
+type BaseProps = {
+  onClose: () => void
+} & Pick<SafeMessage, 'logoUri' | 'name' | 'message'>
+
+// Custom Safe Apps do not have a `safeAppId`
+type ProposeProps = BaseProps & {
+  safeAppId?: number
+  messageHash?: never
+  requestId: RequestId
+}
+
+// A proposed message does not return the `safeAppId` but the `logoUri` and `name` of the Safe App that proposed it
+type ConfirmProps = BaseProps & {
+  safeAppId?: never
+  messageHash: string
+  requestId?: RequestId
+}
+
+const MsgModal = ({
+  onClose,
+  logoUri,
+  name,
+  message,
+  messageHash,
+  safeAppId,
+  requestId,
+}: ProposeProps | ConfirmProps): ReactElement => {
+  // Hooks & variables
+  const [submitError, setSubmitError] = useState<Error | undefined>()
+
+  const web3 = useWeb3()
+  const { safe } = useSafeInfo()
+  const isWrongChain = useIsWrongChain()
+  const isOwner = useIsSafeOwner()
+  const wallet = useWallet()
+  const messages = useSafeMessages()
+
+  // Decode message if UTF-8 encoded
+  const decodedMessage = useMemo(() => {
+    return typeof message === 'string' ? getDecodedMessage(message) : message
+  }, [message])
+
+  // Get `SafeMessage` message
+  const safeMessageMessage = useMemo(() => {
+    return generateSafeMessageMessage(decodedMessage)
+  }, [decodedMessage])
+
+  // Get `SafeMessage` hash
+  const safeMessageHash = useMemo(() => {
+    return messageHash ?? generateSafeMessageHash(safe, decodedMessage)
+  }, [messageHash, safe, decodedMessage])
+
+  // Get already proposed message
+  const [alreadyProposedMessage] = useAsync<SafeMessage | Omit<SafeMessage, 'type'>>(() => {
+    const localMessage = messages.page?.results
+      .filter(isSafeMessageListItem)
+      .find((message) => message.messageHash === messageHash)
+
+    return localMessage ? Promise.resolve(localMessage) : getSafeMessage(safe.chainId, safeMessageHash)
```

Comment:
In my tests, it was always trying to fetch the message and getting a 404 when I was signing a new message. Is this intended?

Reply:
Yes, we need to determine whether to propose or confirm the message. They are different endpoints.